lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+      std::error_code EC;
+      llvm::raw_fd_ostream *fd_stream = new llvm::raw_fd_ostream(
+          SILStatsOutputFile, EC, llvm::sys::fs::OpenFlags::F_Text);
+      if (!fd_stream->has_error() && !EC) {
+        stats_output_stream = {fd_stream,
+                               [](llvm::raw_ostream *d) { delete d; }};
+        return *stats_output_stream.get();
+      }
+      fd_stream->clear_error();
+      llvm::errs() << SILStatsOutputFile << " : " << EC.message() << "\n";
+      delete fd_stream;
+    }
+    // Otherwise use llvm::errs() as output. No need to destroy it at the end.
+    stats_output_stream = {&llvm::errs(), [](llvm::raw_ostream *d) {}};
+  }
+  return *stats_output_stream.get();
+}
+
+/// A helper function to dump the counter value.
+void printCounterValue(StringRef Kind, StringRef CounterName, int CounterValue,
+                       StringRef Symbol, TransformationContext &Ctx) {
+  stats_os() << Kind;
+  stats_os() << ", ";
+
+  stats_os() << CounterName;
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getStageName();
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getTransformId();
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getPassNumber();
+  stats_os() << ", ";
+
+  stats_os() << CounterValue;
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getDuration();
+  stats_os() << ", ";
+
+  stats_os() << Symbol;
+  stats_os() << "\n";
+}
+
+/// A helper function to dump the change of the counter value.
+void printCounterChange(StringRef Kind, StringRef CounterName, double Delta,
+                        int OldValue, int NewValue, TransformationContext &Ctx,
+                        StringRef Symbol = "") {
+  stats_os() << Kind;
+  stats_os() << ", ";
+
+  stats_os() << CounterName;
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getStageName();
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getTransformId();
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getPassNumber();
+  stats_os() << ", ";
+
+  llvm::format_provider<double>::format(Delta, stats_os(), "f8");
+  stats_os() << ", ";
+
+  stats_os() << OldValue;
+  stats_os() << ", ";
+
+  stats_os() << NewValue;
+  stats_os() << ", ";
+
+  stats_os() << Ctx.getDuration();
+  stats_os() << ", ";
+
+  stats_os() << Symbol;
+  stats_os() << "\n";
+}
+
+/// Check if a function name matches the pattern provided by the user.
+bool isMatchingFunction(SILFunction *F, bool shouldHaveNamePattern = false) {
+  auto FuncName = F->getName();
+  // Is it an exact string match?
+  if (!StatsOnlyFunctionName.empty())
+    return F->getName() == StatsOnlyFunctionName;
+
+  // Does the name contain a user-defined substring?
+  if (!StatsOnlyFunctionsNamePattern.empty()) {
+    StringRef Pattern(StatsOnlyFunctionsNamePattern);
+    return FuncName.find(Pattern) != StringRef::npos;
+  }
+
+  return shouldHaveNamePattern ? true : false;
+}
+
+/// Compute the delta between the old and new values.
+/// Return it as a percent value.
+double computeDelta(int Old, int New) {
```

Comment:
static?

Reply:
It is inside an anonymous namespace anyways. So, it cannot be referenced from outside.

Reply:
SGTM.